internal/cloud/backend.go

```diff
@@ -955,43 +955,85 @@ func (b *Cloud) Operation(ctx context.Context, op *backendrun.Operation) (*backe
 
 		defer b.opLock.Unlock()
 
-		r, opErr := f(stopCtx, cancelCtx, op, w)
-		if opErr != nil && opErr != context.Canceled {
-			var diags tfdiags.Diagnostics
-			diags = diags.Append(opErr)
-			op.ReportResult(runningOp, diags)
-			return
-		}
+		// Execute another code path for a query operation
+		// In Terraform a query operation is very similar to a regular plan, but
+		// in HCP Terraform there is a different set of endpoints
+		if op.Query {
+			r, opErr := b.opQuery(stopCtx, cancelCtx, op, w)
+			if opErr != nil && opErr != context.Canceled {
+				var diags tfdiags.Diagnostics
+				diags = diags.Append(opErr)
+				op.ReportResult(runningOp, diags)
+				return
+			}
 
-		if r == nil && opErr == context.Canceled {
-			runningOp.Result = backendrun.OperationFailure
-			return
-		}
+			if r == nil && opErr == context.Canceled {
+				runningOp.Result = backendrun.OperationFailure
+				return
+			}
 
-		if r != nil {
-			// Retrieve the run to get its current status.
-			r, err := b.client.Runs.Read(cancelCtx, r.ID)
```

Comment:
I wonder if we can create a run-like interface that supports either type (tfe.Run, tfe.QueryRun) of operation result in order to avoid all this? Both are readable and cancelable with an errored status...